docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+<img title="" src="assets/04-state-transitions-scaleup-1-n.excalidraw.png" alt="" data-align="center">
+
+#### Restart of a Voting Etcd-Member in a Multi-Node Etcd Cluster
+
+Following DFA represents the states, sub-states and transitions when a voting etcd-member in a multi-node etcd cluster restarts.
+
+<img title="" src="assets/04-state-transitions-restart-member.excalidraw.png" alt="" data-align="center">
+
+> NOTE: If the DB validation fails then data directory of the etcd-member is removed and etcd-member is removed from cluster membership, thus transitioning it to `New` state. The state transitions from `New` state are depicted by [this section](#addition-of-a-new-etcd-member-in-a-multi-node-etcd-cluster).
+
+### Deterministic Etcd Member Creation/Restart During Scale-Up
+
+**Bootstrap information:**
+
+When an etcd-member starts, then it needs to find out:
+
+* If it should join an existing cluster or start a new cluster.
+
+* If it should add itself as a `Learner` or directly start as a voting member.
+
+**Issue with the current approach**:
+
+*At present*, this is facilitated by three things:
+
+1. During scale-up, druid adds an annotation `gardener.cloud/scaled-to-multi-node` to the `StatefulSet`. Each etcd-members looks up this annotation.
+
+2. backup-sidecar attempts to fetch etcd cluster member-list and checks if this etcd-member is already part of the cluster.
+
+3. Size of the cluster by checking `initial-cluster` in the etcd config.
+
+Druid adds an annotation `gardener.cloud/scaled-to-multi-node` on the `StatefulSet` which is then shared by all etcd-members irrespective of the starting state of an etcd-member (as `Learner` or `Voting-Member`). This especially creates an issue for the current leader (often pod with index 0) during the scale-up of an etcd cluster as described in [this](https://github.com/gardener/etcd-backup-restore/issues/646) issue. 
+
+It has been agreed that the current solution to [this]([[BUG] Restart of first cluster member can lead to quorum loss while scaling up a non-HA etcd cluster. · Issue #646 · gardener/etcd-backup-restore · GitHub](https://github.com/gardener/etcd-backup-restore/issues/646)) issue is a quick and dirty fix and needs to be revisited to be uniformly applied to all etcd-members. The authors propose to provide a more deterministic approach to scale-up using the `EtcdMember` resource.
+
+**New approach**
+
+Instead of adding an annotation `gardener.cloud/scaled-to-multi-node` on the `StatefulSet`, a new annotation `druid.gardener.cloud/create-as-learner` should be added by druid on an `EtcdMember` resource. This annotation will only be added to newly created members during scale-up.
+
+Each etcd-member should look at the following to deterministically compute the `bootstrap information` specified above:
+
+* `druid.gardener.cloud/create-as-learner` annotation on its respective `EtcdMember` resource. This new annotation will be honored in the following cases:
```

Comment:
Can you explain why we need this `druid.gardener.cloud/create-as-learner` annotation? From reading the text I got that a single member cluster has one `EtcdMember` resource and when scaling up, one could simply create further `EtcdMember` resources to trigger the scale-up.

Reply:
During a scale-up, new EtcdMember resources will be created, that is correct. However, every backup-sidecar container (which today plays an active role in bootstrapping an etcd peer container), needs to deterministically ascertain two things:

1. Should it create a new cluster or join an existing cluster.
2. Should it join as a learner (when it is started for the first time) or should it directly join as a voting member which could happen either in case of a restart (during scale-up or otherwise) or for a single-member etcd cluster.

Today druid annotates the STS with an annotation which tell any member that a scale-up is currently in progress. This has some issues as this annotation is now shared by all members (even the 0th member). We now propose to remove that annotation and instead define another annotation per `EtcdMember`, which then tells the backup-sidecar how it should bootstrap. We have also mentioned that when the first-ever bootstrap is completed then backup-sidecar will remove this annotation.

Reply:
> Should it create a new cluster or join an existing cluster.

Alternatively, you could have a `spec.clusterId` or a similar spec field that instructs the member to join a existing cluster.

> Should it join as a learner (when it is started for the first time) or should it directly join as a voting member which could happen either in case of a restart (during scale-up or otherwise) or for a single-member etcd cluster.

The text already says that the side-car needs to be able to figure out if it is a member of the cluster, i.e. if it is join as voting member, else join as learner?

Reply:
Druid is going to add the annotation to each member resource as per writeup. How does it know which member is going start? Also, there are too many edge cases in this approach

Reply:
>Also, there are too many edge cases in this approach

@abdasgupta can you list down the edge cases so that it helps us evaluate the current proposal in more light?

> Alternatively, you could have a spec.clusterId or a similar spec field that instructs the member to join a existing cluster.

`spec.clusterId` - yes we can think of this as a possibility. Thanks for the suggestion. Now we have one possibility for using the spec :)

Reply:
Thinking again when the first member is created then the `clusterID` is not known to druid. Therefore it must be part of the `EtcdMember.Status`. Do you think it is wise to include it as part of spec (nil for first member and non-nil for rest of the members) and status?

Reply:
Ok as discussed offline - for the first member `spec.clusterID` will be nil/empty. For all the other members, it will be set by druid. This then removes the need to use any annotation. Every member will also update `status.clusterID`.

Reply:
Discussion amongst: @shreyas-s-rao, @unmarshall and @seshachalam-yv 

@timuthy we can get rid of both the annotation `druid.gardener.cloud/create-as-learner` and the proposed `spec.clusterID`. 
When a member comes up for the first time or is restarted then it should do the following:

> NOTE: Data directory for any member is considered valid if the data is not corrupt. It does not consider revision mismatch with the leader as a criterion for checking validity.

- Check the data directory. 
  - If the data directory is valid and its last state transition capture is either `PendingLearner` or `Learner` then rejoins as `PendingLearner`. If it was `New` then it rejoins as `New`. If it was a voting member then it always rejoins as `Follower` in case of a multi-node cluster and `Leader` if it is a single node etcd cluster.
  - If the data directory is not valid, then anyways data directory will get removed and it will be added as a `PendingLearner` in case of a multi-node cluster. In case of a single-node cluster it will go into `Restoration` from the backup.

The above proposal is only going to work when we make it mandatory to update state/sub-state till the member is learning and is not yet a voting member. Once it becomes a voting member then state transitions (Leader <-> Follower) is done with eventual consistency.

Reply:
Fine for me, thanks 🙂